File: examples/hotp_fn.rs
use rusotp::{generate_hotp, hotp_provisioning_uri, verify_hotp};

fn main() {
    let secret = "12345678901234567890";

    let data = vec![
        (6, 10, 1, "247374"),
        (6, 10, 2, "254785"),
        (6, 10, 3, "496144"),
        (6, 16, 1, "687B4E"),
        (6, 24, 1, "N7C1B6"),
        (6, 36, 1, "M16ONI"),
        (8, 10, 100, "93583477"),
        (8, 16, 100, "23615D75"),
        (8, 24, 100, "032D2EKL"),
        (8, 36, 100, "009TEJXX"),
        (4, 36, 1, "6ONI"),
        (4, 36, 2, "KYWX"),
        (4, 36, 3, "ERBK"),
        (4, 36, 4, "ROTO"),
    ];

    data.iter().for_each(|(length, radix, counter, otp)| {
        if radix == &10 && length == &6 {
            println!(
                "LENGTH: {}, RADIX: {}, COUNTER: {} \tHOTP : {} \tVERIFIED : {} \tURI : {}",
                length,
                radix,
                counter,
                generate_hotp(secret, *length, *radix, *counter),
                verify_hotp(secret, *length, *radix, otp, *counter, 0),
                hotp_provisioning_uri(secret, *length, *radix, "rusotp", *counter)
            );
        } else {
            println!(
                "LENGTH: {}, RADIX: {}, COUNTER: {} \tHOTP : {} \tVERIFIED : {}",
                length,
                radix,
                counter,
                generate_hotp(secret, *length, *radix, *counter),
                verify_hotp(secret, *length, *radix, otp, *counter, 0),
            );
        }
    });
}

File: c_examples/hotp_fn.cpp
#include <stdio.h>
#include "rusotp.h"

int main() {
    const char* secret = "12345678901234567890";

    struct {
        HotpConfig config;
        unsigned long counter;
        const char* otp;
    } data[] = {
        {{"SHA256", secret, 6, 10}, 1, "247374"},
        {{"SHA256", secret, 6, 10}, 2, "254785"},
        {{"SHA256", secret, 6, 10}, 3, "496144"},
        {{"SHA256", secret, 6, 16}, 1, "687B4E"},
        {{"SHA256", secret, 6, 24}, 1, "N7C1B6"},
        {{"SHA256", secret, 6, 36}, 1, "M16ONI"},
        {{"SHA256", secret, 8, 10}, 100, "93583477"},
        {{"SHA256", secret, 8, 16}, 100, "23615D75"},
        {{"SHA256", secret, 8, 24}, 100, "032D2EKL"},
        {{"SHA256", secret, 8, 36}, 100, "009TEJXX"},
        {{"SHA256", secret, 4, 36}, 1, "6ONI"},
        {{"SHA256", secret, 4, 36}, 2, "KYWX"},
        {{"SHA256", secret, 4, 36}, 3, "ERBK"},
        {{"SHA256", secret, 4, 36}, 4, "ROTO"},
    };

    for (int i = 0; i < sizeof(data) / sizeof(data[0]); i++) {
        HotpConfig config = data[i].config;
        unsigned long counter = data[i].counter;
        const char* otp_value = data[i].otp;

        char *otp = generate_hotp(config, counter);
        const char *verified = verify_hotp(config, otp, counter, 0) ? "true" : "false";

        if (config.radix == 10 && config.length == 6 && strcmp(config.algorithm, "SHA1") == 0) {
            char *uri = hotp_provisioning_uri(config, "rusotp", counter);
            printf("LENGTH: %d, RADIX: %d, COUNTER: %lu \tHOTP : %s \tVERIFIED : %s \tURI : %s\n", config.length, config.radix, counter, otp, verified, uri);
        } else {
            printf("LENGTH: %d, RADIX: %d, COUNTER: %lu \tHOTP : %s \tVERIFIED : %s\n", config.length, config.radix, counter, otp, verified);
        }
    }

    return 0;
}